cli/util/upload_bin.go

```diff
@@ -227,33 +227,44 @@ func uploadBlocksAndIndexFiles(ctx *cli.Context, p pool.Pool, rpc *rpcclient.Cli
 }
 
 // searchIndexFile returns the ID and buffer for the next index file to be uploaded.
-func searchIndexFile(ctx *cli.Context, p pool.Pool, containerID cid.ID, privKeys *keys.PrivateKey, signer user.Signer, indexFileSize uint, blockAttributeKey, attributeKey string, maxParallelSearches, maxRetries uint, debug bool) (uint, []byte, error) {
+func searchIndexFile(ctx *cli.Context, p pool.Pool, containerID cid.ID, privKeys *keys.PrivateKey, indexFileSize uint, blockAttributeKey, attributeKey string, maxParallelSearches, maxRetries uint, debug bool) (uint, []byte, error) {
 	var (
 		// buf is used to store OIDs of the uploaded blocks.
 		buf    = make([]byte, indexFileSize*oid.Size)
 		doneCh = make(chan struct{})
 		errCh  = make(chan error)
 
-		existingIndexCount = uint(0)
+		existingIndexCount = 0
 		filters            = object.NewSearchFilters()
 	)
 	go func() {
 		defer close(doneCh)
 		// Search for existing index files.
+		filters.AddFilter(attributeKey, fmt.Sprintf("%d", existingIndexCount), object.MatchNumGE)
 		filters.AddFilter("IndexSize", fmt.Sprintf("%d", indexFileSize), object.MatchStringEqual)
-		for i := 0; ; i++ {
-			indexIDs := searchObjects(ctx.Context, p, containerID, privKeys, attributeKey, uint(i), uint(i+1), 1, maxRetries, debug, errCh, filters)
-			resOIDs := make([]oid.ID, 0, 1)
-			for id := range indexIDs {
-				resOIDs = append(resOIDs, id)
-			}
-			if len(resOIDs) == 0 {
-				break
+
+		var indexObsj []client.SearchResultItem
+		err := retry(func() error {
+			var errBlockSearch error
+			indexObsj, errBlockSearch = neofs.ObjectSearch(ctx.Context, &p, privKeys, containerID.String(), filters, []string{attributeKey})
+			return errBlockSearch
+		}, maxRetries, debug)
+		if err != nil {
+			select {
+			case errCh <- fmt.Errorf("failed to search for index file: %w", err):
+			default:
 			}
-			if len(resOIDs) > 1 {
-				fmt.Fprintf(ctx.App.Writer, "WARN: %d duplicated index files with index %d found: %s\n", len(resOIDs), i, resOIDs)
+			return
+		}
+		if len(indexObsj) != 0 {
+			existingIndexCount, err = strconv.Atoi(indexObsj[len(indexObsj)-1].Attributes[0])
```

Comment:
atoi accepts negatives, i think `strconv.ParseUint(indexObsj[len(indexObsj)-1].Attributes[0], 10, 32)` should be used instead, here and below